content/Watsonx/WatsonxAI/files/rag-with-discovery.ipynb

```diff
@@ -0,0 +1,295 @@
+    "        disc_results = []\n",
+    "        for item in passages:\n",
+    "            item = item[\"passage_text\"].replace(\"<em>\",\"\")\n",
+    "            item = item.replace(\"</em>\", \"\")\n",
+    "            disc_results.append(item)\n",
+    "    combined_disc_results.append(\"\\n\".join(disc_results))\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 50,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# LLM that we want to use with watsonx.ai\n",
+    "model_id= \"google/flan-ul2\"\n",
+    "\n",
+    "endpoint= \"https://us-south.ml.cloud.ibm.com\"\n",
+    "\n",
+    "access_token = ''\n",
+    "\n",
+    "try:\n",
+    "  access_token = IAMTokenManager(\n",
+    "    apikey = api_key,\n",
+    "    url = \"https://iam.cloud.ibm.com/identity/token\"\n",
+    "  ).get_token()\n",
+    "except:\n",
+    "  print('Issue obtaining access token. Check variables?')\n",
+    "\n",
+    "credentials = { \n",
+    "    \"url\"    : endpoint, \n",
+    "    \"token\" : access_token\n",
+    "}\n",
+    "\n",
+    "# Project ID of your watsonx project\n",
+    "watsonx_project_id = \"INSERT YOUR watsonx PROJECT ID HERE\"\n",
+    "\n",
+    "# watsonx.ai tuning parameters\n",
+    "gen_params = {\n",
+    "    \"DECODING_METHOD\" : \"greedy\",\n",
+    "    \"MAX_NEW_TOKENS\" : 500,\n",
+    "    \"MIN_NEW_TOKENS\" : 20,\n",
+    "    \"STREAM\" : False,\n",
+    "    \"TEMPERATURE\" : 1.7,\n",
+    "    \"TOP_K\" : 50,\n",
+    "    \"TOP_P\" : 1,\n",
+    "    \"RANDOM_SEED\" : 10\n",
+    "}\n",
+    "\n",
+    "model = Model( model_id, credentials, gen_params, watsonx_project_id )"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 51,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\n",
+      "Answer:\n",
+      "€2,250 million.'] ### As of March 31, 2023, the Company had €2,250 million of outstanding long- term debt and approximately €42 million of accrued interest designated as non-derivative hedges of its net investment in certain foreign operations.\n"
+     ]
+    }
+   ],
+   "source": [
+    "#######################################################################################\n",
+    "prompt_template = \"\"\"\n",
+    "Article:\n",
+    "###\n",
+    "%s\n",
+    "###\n",
+    "\n",
+    "Answer the following question using only information from the article. \n",
+    "Answer in a complete sentence, with proper capitalization and punctuation. \n",
+    "If there is no good answer in the article, say \"I don't know\".\n",
+    "\n",
+    "Question: %s\n",
+    "Answer: \n",
+    "\"\"\"\n",
+    "\n",
+    "#######################################################################################\n",
+    "def augment( template_in, context_in, query_in ):\n",
+    "    return template_in % ( context_in,  query_in )\n",
+    "\n",
```

Comment:
Can you pass in `content_in` in the augment function as `context_in[0]`? This will remove the `']` from the final answer:
<img width="470" alt="Screenshot 2023-10-03 at 12 29 58 PM" src="https://github.com/ibm-build-lab/VAD-VAR-Workshop/assets/19367627/796dab67-3a2e-4a9e-b8ec-9abc58be62a9">

Reply:
Mentioned this in the other [reply](https://github.com/ibm-build-lab/VAD-VAR-Workshop/pull/196#discussion_r1344616080), but the extra content is result of extra tokens that the LLM is trying to fill in, reducing that parameter addressed the extra tokens.